cli/internal/taskhash/taskhash.go

```diff
@@ -277,18 +277,68 @@ func (th *Tracker) CalculateFileHashes(
 	return nil
 }
 
-type taskHashInputs struct {
+type taskHashable struct {
 	packageDir           turbopath.AnchoredUnixPath
 	hashOfFiles          string
 	externalDepsHash     string
 	task                 string
 	outputs              fs.TaskOutputs
 	passThruArgs         []string
+	envMode              util.EnvMode
+	passthroughEnv       []string
 	hashableEnvPairs     []string
 	globalHash           string
 	taskDependencyHashes []string
 }
 
+type oldTaskHashable struct {
+	packageDir           turbopath.AnchoredUnixPath
+	hashOfFiles          string
+	externalDepsHash     string
+	task                 string
+	outputs              fs.TaskOutputs
+	passThruArgs         []string
+	hashableEnvPairs     []string
+	globalHash           string
+	taskDependencyHashes []string
+}
+
+// calculateTaskHashFromHashable returns a hash string from the taskHashable
+func calculateTaskHashFromHashable(full *taskHashable) (string, error) {
+	switch full.envMode {
+	case util.Infer:
+		if full.passthroughEnv != nil {
+			// In infer mode, if there is any passThru config (even if it is an empty array)
+			// we'll hash the whole object, so we can detect changes to that config
+			// Further, resolve the envMode to the concrete value.
+			full.envMode = util.Strict
+			return fs.HashObject(full)
+		}
+
+		// If we're in infer mode, and there is no global pass through config,
+		// we can use the old anonymous struct. this will be true for everyone not using the strict env
+		// feature, and we don't want to break their cache.
+		return fs.HashObject(&oldTaskHashable{
+			packageDir:           full.packageDir,
+			hashOfFiles:          full.hashOfFiles,
+			externalDepsHash:     full.externalDepsHash,
+			task:                 full.task,
+			outputs:              full.outputs,
+			passThruArgs:         full.passThruArgs,
+			hashableEnvPairs:     full.hashableEnvPairs,
+			globalHash:           full.globalHash,
+			taskDependencyHashes: full.taskDependencyHashes,
+		})
+	case util.Loose:
+		// Remove the passthroughs from hash consideration if we're explicitly loose.
+		full.passthroughEnv = nil
+		return fs.HashObject(full)
+	default:
+		// When we aren't in infer or loose mode we can hash the whole object as is.
+		return fs.HashObject(full)
+	}
+}
```

Comment:
Is this necessary? I can understand avoiding busting a global hash between releases, but seems overkill to avoid busting individual task hashes too. CAn we just `fs.HashObject(full)` for all these modes?

Reply:
Hashing `full` here is effectively a global hash change since it applies to _all_ tasks by changing the object shape.

Reply:
Also, this is still "experimental" so if we start hashing it in we're committing to that hash shape for everybody before stabilizing it.